nucleus/common/internal-api/src/main/java/org/glassfish/internal/api/Globals.java

```diff
@@ -144,7 +154,50 @@ public static ServiceLocator getStaticHabitat() {
 
         return defaultHabitat;
     }
+
+    /**
+     * Returns the static service locator, but using the classloader used to load
+     * the modules/ directory. This is useful when you need access to classes from
+     * the modules/ directory from client code.
+     * 
+     * @return the static service locator with modules loaded
+     * @throws MalformedURLException if the URL to a module is invalid
+     */
+    public static ServiceLocator getStaticHabitatWithModules() throws MalformedURLException {
+        if (defaultHabitatWithModules == null) {
+            synchronized(STATIC_LOCK) {
+                if (defaultHabitatWithModules == null) {
+                    defaultHabitatWithModules = createStaticHabitatWithModules();
+                }
+            }
+        }
+        return defaultHabitatWithModules;
+    }
     
+    private static ServiceLocator createStaticHabitatWithModules() throws MalformedURLException {
+        // get the list of JAR files from the modules directory
+        ArrayList<URL> urls = new ArrayList<>();
+        File idir = new File(System.getProperty(SystemPropertyConstants.INSTALL_ROOT_PROPERTY));
+        File mdir = new File(idir, "modules");
+        for (File f : mdir.listFiles()) {
+            if (f.toString().endsWith(".jar")) {
+                urls.add(f.toURI().toURL());
+            }
+        }
+
+        final URL[] urlsA = urls.toArray(new URL[urls.size()]);
+
+        ClassLoader cl = (ClassLoader) AccessController.doPrivileged(new PrivilegedAction<Object>() {
```

Comment:
Set the action to the variable -> you don't need any casting then, and also you don't need an anonymous class, you can use lambda.
```
PrivilegedAction<URLClassLoader> action = () -> {...};
ClassLoader cl = AccessController.doPrivileged(action);
```

Reply:
Also ... I will have conflicts in this class, but that's my problem ;-)

Reply:
You're right sorry, I'd been working on Payara 4 so I'd been in the mindset of not using Java 8 features 😆